source/verus/src/main.rs

```diff
@@ -75,19 +51,263 @@ fn main() {
+        "\n{} {}",
+        yansi::Paint::blue("removing:"),
+        std::env::current_dir().unwrap().join(dep_file_path.clone()).display()
+    );
+    std::fs::remove_file(dep_file_path).expect("remove crate root dependency file");
+
+    // step 5, git commiting
+    println!("\n{} {}", yansi::Paint::blue("commiting"), proj_path.display());
+    std::process::Command::new("git")
+        .current_dir(&proj_path)
+        .arg("add")
+        .args(deps)
+        .arg("reports.toml")
+        .output()
+        .map_err(|x| format!("Telemetry: failed to git add with error message: {}", x))?;
+
+    std::process::Command::new("git")
+        .current_dir(&proj_path)
+        .arg("commit")
+        .arg("-m")
+        .arg("\"verus telemtry\"")
+        .output()
+        .map_err(|x| format!("Telemetry: failed to git commit with error message: {}", x))?;
+
+    Ok(())
+}
+
+pub type ReportsPath = Option<Reports>;
+#[derive(Debug)]
+pub struct Reports {
+    proj_path: PathBuf,
+    crate_root: PathBuf,
+}
+
+// is it easier to impl the copy trait? I'm not sure about the paradigm here
+impl Clone for Reports {
+    fn clone(&self) -> Self {
+        Self { proj_path: self.proj_path.clone(), crate_root: self.crate_root.clone() }
+    }
+}
+
+fn repo_path() -> ReportsPath {
+    // Step 1: check if there's a verus/reports/.git file in the user's local data lib
+    // if not so, create verus/ directory
+    let cache_dir = dirs::data_local_dir()?;
+    let reports_dir = cache_dir.join("verus").join("reports");
+    if !reports_dir.clone().join(".git").is_file() {
+        std::fs::create_dir_all(reports_dir.clone())
+            .expect("Creating verus/ directory in local data cache");
+
+        std::process::Command::new("git")
+            .current_dir(&reports_dir)
+            .arg("init")
+            .output()
+            .expect("Initializing git in local data cache");
+    }
+
+    let uuid_file = reports_dir.join("userid");
+    if !uuid_file.is_file() {
+        let uuid = uuid::Uuid::new_v4();
+        std::fs::write(uuid_file.clone(), uuid.to_string()).expect("Writing user id file");
+    }
+
+    let uuid = std::fs::read_to_string(uuid_file).expect("Reading user id file");
+
+    // Step 3: project dir check/create (project = SHA256((normalized/absolute)path to root.rs)
+
+    // clone is not implemented for Args, so I grabbed the arguments here again
+    let new_args = std::env::args().into_iter();
+
+    let rs_files = new_args.filter(|arg| arg.ends_with(".rs")).collect::<Vec<_>>();
+
+    let input_file: Option<String> = match rs_files.len() {
+        1 => Some(rs_files[0].clone()),
+        _ => {
+            eprintln!("{}: cannot find input file", yansi::Paint::red("error"));
+            None // passed down to rust_verify for error message
+        }
+    };
+
+    let input_file_path = match input_file.clone() {
+        Some(input_file_name) => std::fs::canonicalize(input_file_name).ok(),
+        None => None,
+    };
```

Comment:
You can likely write this more cleanly using `Option::and_then` (eg: `input_file.as_ref().and_then(|f| std::fs::canonicalize(f).ok())`), also `Path` has a [`canonicalize` method](https://doc.rust-lang.org/std/path/struct.Path.html#method.canonicalize), so you can just use `input_file.as_ref().and_then(|f| f.canonicalize().ok())`